// @sap/grunt-sapui5-bestpractice-build/Gruntfile.js
module.exports = function(grunt) {
    "use strict";

    var config = {

        mochaTest: {
            test: {
                options: {
                    reporter: 'mocha-junit-reporter',
                    reporterOptions: {
                        mochaFile: 'testResults/TEST-index.xml'
                    },
                    colors: true,
                    summery: true,
                    captureFile: 'testResults/output.txt', // Optionally capture the reporter output to a file
                    quiet: false, // Optionally suppress output to standard out (defaults to false)
                    clearRequireCache: false // Optionally clear the require cache before running tests (defaults to false)
                },
                src: ['test/*Spec.js']
            }
        },
        eslint: {
            src: [
                "tasks/**/*.js"
            ],
            options: {
                configFile: "cfg/eslint.json",
                outputFile: "target/eslint.xml"
            }
        },
        auto_install: {
            mtaDummy: {
                options: {
                    cwd: './test/dummy/mtaDummy/CrudModule',
                    stderr: false,
                    failOnError: true
                }
            },
            rootProject: {
                options: {
                    cwd: './test/dummy/rootProjectOldStructure',
                    stderr: false,
                    failOnError: true
                }
            },
            webcontent: {
                options: {
                    cwd: './test/dummy/webcontentOldStructure',
                    stderr: false,
                    failOnError: true
                }
            },
            webapp: {
                options: {
                    cwd: './test/dummy/webappNewStructure',
                    stderr: false,
                    failOnError: true
                }
            }
        }
    };

    grunt.config.merge(config);

    grunt.loadNpmTasks('grunt-auto-install');
    grunt.loadNpmTasks('grunt-mocha-test');
    grunt.loadNpmTasks("gruntify-eslint");

    grunt.registerTask('validations', [
        'eslint'
    ]);

    grunt.registerTask('devx-grunt-build-voter', [
        // 'auto_install:rootProject',
        // 'auto_install:webcontent',
        'auto_install:webapp',
        'auto_install:mtaDummy',
        'mochaTest'
    ]);

    grunt.registerTask('default', [
        'devx-grunt-build-voter',
        'validations'
    ]);
};
